File: src/main/resources/static/rsaService.js
export class RsaService {
    static generateTwoDifferentPrimesWithExplanation() {
        let p, q;
        do {
            p = this.generateTwoDigitPrime();
            q = this.generateTwoDigitPrime();
        } while (p === q || !this.hasValidE(this.phiWithExplanation(BigInt(p), BigInt(q)).phi));

        return { p, q };
    }

    static generateTwoDigitPrime() {
        let prime;
        do {
            prime = Math.floor(Math.random() * (99 - 10 + 1)) + 10;
        } while (!this.isProbablePrime(prime));
        return prime;
    }

    static isProbablePrime(num) {
        if (num <= 1) return false;
        if (num <= 3) return true;
        if (num % 2 === 0 || num % 3 === 0) return false;

        for (let i = 5; i * i <= num; i += 6) {
            if (num % i === 0 || num % (i + 2) === 0) return false;
        }
        return true;
    }

    static phiWithExplanation(p, q) {
        const phi = (p - 1n) * (q - 1n);
        const explanation = `phi(n) = (p - 1) * (q - 1) = (${p} - 1) * (${q} - 1) = ${phi}`;
        return { phi, explanation };
    }

    static hasValidE(phi) {
        for (let e = 3n; e <= 5n; e++) {
            if (this.gcd(e, phi) === 1n) {
                return true;
            }
        }
        return false;
    }

    static selectEWithExplanation(phi) {
        const explanation = [];

        const quersumme = phi.toString().split('').reduce((acc, digit) => acc + BigInt(digit), 0n);
        if (quersumme % 3n !== 0n) {
            explanation.push(`e = 3 ist gültig, da die Quersumme von ${phi} = ${quersumme} nicht durch 3 teilbar ist.`);
            return { e: 3n, explanation: explanation.join("\n") };
        } else {
            explanation.push(`e = 3 ist nicht gültig, da die Quersumme von ${phi} = ${quersumme} durch 3 teilbar ist.`);
        }

        const lastTwoDigits = phi % 100n;
        if (lastTwoDigits % 4n !== 0n) {
            explanation.push(`e = 4 ist gültig, da die letzten beiden Ziffern=${lastTwoDigits} nicht durch 4 teilbar ist.`);
            return { e: 4n, explanation: explanation.join("\n") };
        } else {
            explanation.push(`e = 4 ist nicht gültig, da die letzten beiden Ziffern=${lastTwoDigits} durch 4 teilbar ist.`);
        }

        const lastDigit = phi % 10n;
        if (lastDigit !== 5n && lastDigit !== 0n) {
            explanation.push(`e = 5 ist gültig, da die letzte Stelle von ${phi} = ${lastDigit} nicht 5 oder 0 ist.`);
            return { e: 5n, explanation: explanation.join("\n") };
        } else {
            explanation.push(`e = 5 ist nicht gültig, da die letzte Stelle von ${phi} = ${lastDigit} 5 oder 0 ist.`);
        }

        explanation.push("Kein gültiges e gefunden!");
        return { e: null, explanation: explanation.join("\n") };
    }

    static gcd(a, b) {
        while (b !== 0n) {
            const temp = b;
            b = a % b;
            a = temp;
        }
        return a;
    }

    static encryptWithExplanation(m, e, n) {
        const c = this.modPow(m, e, n);
        const explanation = `Verschlüsselung: c = m^e mod n = ${m}^${e} mod ${n} = ${c}`;
        return { c, explanation };
    }

    static decryptWithExplanation(c, d, n) {
        const m = this.modPow(c, d, n);
        const explanation = `Entschlüsselung: m = c^d mod n = ${c}^${d} mod ${n} = ${m}`;
        return { m, explanation };
    }

    static modPow(base, exp, mod) {
        let result = 1n;
        base = base % mod;

        while (exp > 0n) {
            if (exp % 2n === 1n) {
                result = (result * base) % mod;
            }
            exp = exp / 2n;
            base = (base * base) % mod;
        }

        return result;
    }
}

File: src/main/kotlin/cc/sokolowski/informationssicherheit/rsa/RsaService.kt
package cc.sokolowski.informationssicherheit.rsa

import java.math.BigInteger
import java.security.SecureRandom
import org.springframework.stereotype.Service

@Service
class RsaService {

    fun generateTwoDifferentPrimesWithExplanation(random: SecureRandom): Triple<BigInteger, BigInteger, String> {
        var p: BigInteger
        var q: BigInteger
        val explanation = StringBuilder("Generierung zufälliger Primzahlen:\n")

        do {
            p = generateTwoDigitPrime(random)
            q = generateTwoDigitPrime(random)
        } while (p == q || !hasValidE(phiWithExplanation(p, q).first))

        explanation.append("p = $p\nq = $q")

        return Triple(p, q, explanation.toString())
    }

    private fun generateTwoDigitPrime(random: SecureRandom): BigInteger {
        var prime: BigInteger
        do {
            prime = BigInteger.valueOf(10 + random.nextInt(90).toLong()) // Bereich 10-99
        } while (!prime.isProbablePrime(100))
        return prime
    }

    private fun hasValidE(phi: BigInteger): Boolean {
        for (e in 3..5) {
            if (phi.gcd(BigInteger.valueOf(e.toLong())) == BigInteger.ONE) {
                return true
            }
        }
        return false
    }

    fun phiWithExplanation(p: BigInteger, q: BigInteger): Pair<BigInteger, String> {
        val phi = (p - BigInteger.ONE) * (q - BigInteger.ONE)
        val explanation = "phi(n) = (p - 1) * (q - 1) = ($p - 1) * ($q - 1) = ${p - BigInteger.ONE} * ${q - BigInteger.ONE} = $phi"
        return Pair(phi, explanation)
    }

    fun selectE(phi: BigInteger): Pair<BigInteger, String> {
        val steps = StringBuilder("Schritte zur Auswahl von e:\n")

        val quersumme = phi.toString().map { it.toString().toInt() }.sum()
        if (quersumme % 3 != 0) {
            steps.append("e = 3 ist gültig, da die Quersumme von $phi = $quersumme nicht durch 3 teilbar ist.\n")
            return Pair(BigInteger.valueOf(3), steps.toString())
        } else {
            steps.append("e = 3 ist nicht gültig, da die Quersumme von $phi = $quersumme durch 3 teilbar ist.\n")
        }

        val lastTwoDigits = phi.mod(BigInteger.valueOf(100))
        if (lastTwoDigits % BigInteger.valueOf(4) != BigInteger.ZERO) {
            steps.append("e = 4 ist gültig, da $lastTwoDigits nicht durch 4 teilbar ist.\n")
            return Pair(BigInteger.valueOf(4), steps.toString())
        } else {
            steps.append("e = 4 ist nicht gültig, da $lastTwoDigits durch 4 teilbar ist.\n")
        }

        val lastDigit = phi.mod(BigInteger.TEN)
        if (lastDigit != BigInteger.valueOf(5) && lastDigit != BigInteger.ZERO) {
            steps.append("e = 5 ist gültig, da die letzte Stelle von $phi = $lastDigit nicht 5 oder 0 ist.\n")
            return Pair(BigInteger.valueOf(5), steps.toString())
        } else {
            steps.append("e = 5 ist nicht gültig, da die letzte Stelle von $phi = $lastDigit 5 oder 0 ist.\n")
        }

        return Pair(BigInteger.ZERO, steps.append("Kein gültiges e gefunden!\n").toString())
    }

    fun encryptWithExplanation(m: BigInteger, e: BigInteger, n: BigInteger): Pair<BigInteger, String> {
        val c = m.modPow(e, n)
        val explanation = "Verschlüsselung: c = m^e mod n = $m^$e mod $n = $c"
        return Pair(c, explanation)
    }

    fun decryptWithExplanation(c: BigInteger, d: BigInteger, n: BigInteger): Pair<BigInteger, String> {
        val m = c.modPow(d, n)
        val explanation = "Entschlüsselung: m = c^d mod n = $c^$d mod $n = $m"
        return Pair(m, explanation)
    }
}
